sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/DecorrelateInnerQuery.scala

```diff
@@ -428,7 +451,132 @@ object DecorrelateInnerQuery extends PredicateHelper {
+            // | a | b |     | c | d |     | c | cnt |
+            // +---+---+     +---+---+     +---+-----+
+            // | 0 | 1 |     | 0 | 2 |     | 0 | 2   |
+            // | 1 | 2 |     | 0 | 3 |     +---+-----+
+            // +---+---+     +---+---+
+            //
+            // T1 nested loop join T2     T1 left outer join T2'
+            // on (a = c):                on (a = c):
+            // +---+-----+                +---+-----++
+            // | a | cnt |                | a | cnt  |
+            // +---+-----+                +---+------+
+            // | 0 | 2   |                | 0 | 2    |
+            // | 1 | 0   | <--- correct   | 1 | null | <--- wrong result
+            // +---+-----+                +---+------+
+            //
+            // If an aggregate is subject to the COUNT bug:
+            // 1) add a column `true AS alwaysTrue` to the result of the aggregate
+            // 2) insert a left outer domain join between the outer query and this aggregate
+            // 3) rewrite the original aggregate's output column using the default value of the
+            //    aggregate function and the alwaysTrue column.
+            //
+            // For example, T1 left outer join T2' with `alwaysTrue` marker:
+            // +---+------+------------+--------------------------------+
+            // | c | cnt  | alwaysTrue | if(isnull(alwaysTrue), 0, cnt) |
+            // +---+------+------------+--------------------------------+
+            // | 0 | 2    | true       | 2                              |
+            // | 0 | null | null       | 0                              |  <--- correct result
+            // +---+------+------------+--------------------------------+
+            if (groupingExpressions.isEmpty && handleCountBug) {
+              // Evaluate the aggregate expressions with zero tuples.
+              val resultMap = RewriteCorrelatedScalarSubquery.evalAggregateOnZeroTups(newAggregate)
+              val alwaysTrue = Alias(Literal.TrueLiteral, "alwaysTrue")()
+              val alwaysTrueRef = alwaysTrue.toAttribute.withNullability(true)
+              val expressions = ArrayBuffer.empty[NamedExpression]
+              // Create new aliases for aggregate expressions that have non-null default
+              // values and reconstruct the output with the `alwaysTrue` marker.
+              val projectList = newAggregate.aggregateExpressions.map { a =>
+                resultMap.get(a.exprId) match {
+                  // Aggregate expression is not subject to the count bug.
+                  case Some(Literal(null, _)) | None =>
+                    expressions += a
+                    // The attribute is nullable since it is from the right-hand side of a
+                    // left outer join.
+                    a.toAttribute.withNullability(true)
+                  case Some(default) =>
+                    val newAttr = a.newInstance()
```

Comment:
what if we don't call `.newInstance()`?

Reply:
Because we are reusing the `a`'s exprId when constructing the alias, we would have the same `exprId` referencing two different expressions if not using `.newInstance()`. Another option is to use a new exprId for the Alias, but it requires updating parent nodes.
```
Alias(If(IsNull(alwaysTrueRef), default, ref), a.name)(a.exprId)
```

Reply:
Got it.

I think the assumption here is `a` must be an alias. If `a` is an attribute, then `a.newInstance()` will create a new attribute that refers to nothing and breaks the query. Can we add an assert?
```
assert(a.isInstanceOf[Alias])
val newAlias = a.newInstance()
...
```

Reply:
Yes this holds because if `a` is an attribute, evaluating it with zero tuples will yield null:
https://github.com/apache/spark/blob/c46342e3d057bcc949b4caf016514ff05e0a1ebd/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala#L407
Another possibility is OuterReference. Let me verify if outer works in this case.